File: anki_card_maker/lib/latest_file_finder.rb

# Finds the latest modified file in a directory and its subdirectories.
class LatestFileFinder


  def initialize(root_path, file_mask = '*.txt')
    @root_path = root_path
    @file_mask = file_mask
    @last_modified_file = nil
    @last_modified_filedate = nil
  end


  def find
    recurse_find @root_path
    puts "Last modified date/time: #{@last_modified_filedate}"
    @last_modified_file
  end


  def recurse_find(path)
    Dir[File.join(path, @file_mask)].each do |filename|
      if @last_modified_file.nil? or File.mtime(filename) > @last_modified_filedate
        @last_modified_file = filename
        @last_modified_filedate = File.mtime(filename)
      end
    end

    dirs = list_dir(path)
    dirs.each do |dirname|
      recurse_find(File.join(path, dirname))
    end
  end

  def list_dir(path)
    dirs = Dir.entries(path).select do |entry| 
      File.directory? File.join(path, entry) and !(entry =='.' || entry == '..') 
    end
  end

  private :recurse_find


end

# path = '/Users/royce/Dropbox/Documents/Reviewer'
# puts LatestFileFinder.new(path).find

File: Anki Source Analyzer/src/main/java/ph/rye/anki/util/LatestFileFinder.java
/**
 *   Copyright 2016 Royce Remulla
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package ph.rye.anki.util;

import java.io.File;
import java.io.FileFilter;
import java.text.SimpleDateFormat;
import java.util.Locale;

/**
 * @author royce
 *
 */
public class LatestFileFinder {

    private transient File rootPath;
    private transient String fileMask = "*.txt";
    private transient File lastModifiedFile;
    private transient File lastModifiedFolder;
    private transient long lastModifiedFiledate;


    public LatestFileFinder(final String rootPath) {
        this(rootPath, "*.txt");
    }

    public LatestFileFinder(final String rootPath, final String fileMask) {
        this.rootPath = new File(rootPath);
        this.fileMask = fileMask.replace("?", ".?").replace("*", ".*?");
    }


    public File find() {
        recurseFind(rootPath);

        System.out.println(
            String.format(
                "Last modified date/time: %s",
                new SimpleDateFormat("MM/dd/yyyy HH:mm", Locale.getDefault()),
                lastModifiedFiledate));

        return lastModifiedFile;
    }

    private void recurseFind(final File folder) {

        for (final File file : folder.listFiles(
            (FileFilter) pathname -> pathname.isFile()
                    && pathname.getAbsolutePath().matches(fileMask))) {
            if (lastModifiedFile == null
                    || file.lastModified() > lastModifiedFiledate) {
                lastModifiedFile = file;
                lastModifiedFolder = folder;
                lastModifiedFiledate = file.lastModified();
            }
        }

        for (final File subfolder : listFolder(folder)) {
            recurseFind(subfolder);
        }
    }

    private File[] listFolder(final File folder) {
        return folder.listFiles(
            (FileFilter) pathname -> pathname.isDirectory()
                    && !pathname.getName().equals("..")
                    && !pathname.getName().equals("."));
    }

    public File getLatestFolder() {
        return lastModifiedFolder;
    }

}


//# path = '/Users/royce/Dropbox/Documents/Reviewer'
//# puts LatestFileFinder.new(path).find
